Fix Vulkan compiling issues on Windows

mantaionut opened the pull request:
PR based on #61431
Fix USE_VULKAN=1 and USE_VULKAN_WRAPPER=0 not compiling on Windows.
Change designated initializers since they require C++20.
Rename Hasher typename since it's not compiling due to https://developercommunity.visualstudio.com/t/1397858

Fixes #59519

mantaionut commented:
@malfet can you please review these changes?
Thanks.

ezyang commented on aten/src/ATen/native/vulkan/api/Descriptor.cpp:
```diff
@@ -41,20 +41,14 @@ DescriptorSet& DescriptorSet::operator=(DescriptorSet&& other) noexcept {
+  binder.resource_info.buffer_info.buffer = buffer.handle(); // buffer
+  binder.resource_info.buffer_info.offset = buffer.mem_offset(); // offset
+  binder.resource_info.buffer_info.range = buffer.mem_range(); // range
+  add_binding(binder);
```
Consider std::move() here?

mantaionut committed:
Fix Vulkan compiling issues on Windows

Change designated initializers since they require c++20.
Rename Hasher typename since it's not compiling due to
https://developercommunity.visualstudio.com/t/1397858

mantaionut commented:
@ezyang I made the change you requested. Do you know who can review my changes?

SS-JIA commented on c10/util/flat_hash_map.h:
```diff
@@ -234,12 +234,14 @@ template <
     typename FindKey,
     typename ArgumentHash,
-    typename Hasher,
+    typename DetailHasher,
```
Is this change necessary? Seems that it's simply changing the name of `Hasher` to `DetailHasher`. I believe this file was copied over from [this repo](https://github.com/skarupke/flat_hash_map) so I think we should avoid making changes unless necessary.

malfet replied:
It's mentioned in PR description, isn't it
> Rename Hasher typename since it's not compiling due to https://developercommunity.visualstudio.com/t/1397858

SS-JIA reviewed (dismissed):
Overall looks good! But I would suggest removing the changes in `c10/util/flat_hash_map.h` since it doesn't seem to be needed.

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge
